### PR TITLE
List and get terraform and capi templates

### PR DESCRIPTION
**Pull request by Skarlso**
Closes https://github.com/weaveworks/weave-gitops-enterprise/issues/586

## List tf-controller Templates

This change adds the ability to list tf-controller templates which are marked as GitOpsTemplates for the moment.

There are some small refactorings in this PR as well.

- Eliminating the `capi` package and putting the template specific code into the `templates` package.
- Adding the `terraform` command which is the same as the templates command but since we would like to have the discoverable `get template terraform` subcommand, for now the code is a bit duplicated -> this will later be reconciled as we merge some code paths

### CODE REVIEW

File: pkg/templates/templates.go [hunk 1]
```diff
@@ -50,3 +56,245 @@ func CreatePullRequestFromTemplate(params CreatePullRequestFromTemplateParams, r
 
 	return nil
 }
+
```
Comment (Skarlso) [inline pkg/templates/templates.go:59]: All of this is moved code. The only change is adding the `kind` type to template listing.

File: pkg/templates/templates_test.go [hunk 2]
```diff
@@ -111,8 +114,486 @@ func TestCreatePullRequestFromTemplate_Terraform(t *testing.T) {
 
 			c, err := adapters.NewHttpClient(testutils.BaseURI, "", "", client, os.Stdout)
 			assert.NoError(t, err)
-			result, err := c.CreatePullRequestFromTemplate(templates.CreatePullRequestFromTemplateParams{TemplateKind: templates.GitOpsTemplateKind})
+			result, err := c.CreatePullRequestFromTemplate(templates.CreatePullRequestFromTemplateParams{TemplateKind: templates.GitOpsTemplateKind.String()})
 			tt.assertFunc(t, result, err)
 		})
 	}
 }
+
```
Comment (Skarlso) [inline pkg/templates/templates_test.go:122]: Todo: 
- [ ] Add test cases for GitOpsTemplate Kind

**Comment by Skarlso** [comment]
```
./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password=<password> --insecure-skip-tls-verify get templates terraform
NAME                                PROVIDER   DESCRIPTION                                                                                     ERROR
sample-wge-tf-controller-template              This is a sample WGE template that will be translated into a tf-controller specific template.
```

**Comment by Skarlso** [comment]
Fixing various issues with the endpoint declaration requiring positional arguments for getting templates.
https://github.com/weaveworks/weave-gitops-enterprise/pull/855

I wonder if it would be better to remake the getter into POST requests and post the body. That way, the endpoint could remain the same and not be so subject to breakage. 🤔

**Comment by Skarlso** [comment]
Manual test flow:

```
➜  weave-gitops git:(list-tf-templates) ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform
NAME                                PROVIDER   DESCRIPTION                                                                                     ERROR
new-template-2                                 Slightly shorter description
sample-wge-tf-controller-template              This is a sample WGE template that will be translated into a tf-controller specific template.
➜  weave-gitops git:(list-tf-templates) ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform --list-parameters new-template-2
NAME            REQUIRED   DESCRIPTION     OPTIONS
RESOURCE_NAME   false      Resource Name
➜  weave-gitops git:(list-tf-templates) ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template
NAME               PROVIDER   DESCRIPTION               ERROR
cluster-template              this is test template 1
➜  weave-gitops git:(list-tf-templates) ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template --list-parameters cluster-template
NAME           REQUIRED   DESCRIPTION                            OPTIONS
CLUSTER_NAME   false      This is used for the cluster naming.
➜  weave-gitops git:(list-tf-templates) ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform --list-parameters cluster-template
Server error: {"code":2,"message":"error looking up template cluster-template: error getting gitops template default/cluster-template: gitopstemplates.clustertemplates.weave.works \"cluster-template\" not found","details":[]}
Error: unable to retrieve parameters for template "cluster-template" from "https://localhost:8000": response status for GET "https://localhost:8000/v1/templates/cluster-template/GitOpsTemplate/params" was 500
```


I was surprised that there is no single GetTempate by providing a name and that retrieves that single template. ¯\\\_(ツ)\_/¯

**Comment by Skarlso** [comment]
```
./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform new-template-2
NAME             PROVIDER   DESCRIPTION                    ERROR
new-template-2              Slightly shorter description
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template cluster-template
NAME               PROVIDER   DESCRIPTION               ERROR
cluster-template              this is test template 1
```

**Review by chanwit — APPROVED** [review]
LGTM
thank you so much @Skarlso :1st_place_medal:

**Review by jpellizzari — APPROVED** [review]
LGTM. I don't know what the role of the Core CLI is in the future of the product, so this seems odd, but the code looks good.

**Review by yitsushi — APPROVED** [review]

**Review by Himangini — APPROVED** [review]
👍🏻

**Comment by Skarlso** [comment]
Tested without the other PR in enterprise by fixing the parameters vs QueryParams and PathParams. 🤦 

```
./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform
NAME                                PROVIDER   DESCRIPTION                                                                                     ERROR
new-template-2                                 Slightly shorter description
sample-wge-tf-controller-template              This is a sample WGE template that will be translated into a tf-controller specific template.
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform new-template-2
NAME             PROVIDER   DESCRIPTION                    ERROR
new-template-2              Slightly shorter description
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template
NAME               PROVIDER   DESCRIPTION               ERROR
cluster-template              this is test template 1
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template cluster-template
NAME               PROVIDER   DESCRIPTION               ERROR
cluster-template              this is test template 1
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template terraform new-template-2 --list-parameters
NAME            REQUIRED   DESCRIPTION     OPTIONS
RESOURCE_NAME   false      Resource Name
➜  weave-gitops git:(list-tf-templates) ✗ ./bin/gitops --endpoint https://localhost:8000 --username wego-admin --password="super secret password" --insecure-skip-tls-verify get template cluster-template --list-parameters
NAME           REQUIRED   DESCRIPTION                            OPTIONS
```

Works with current enterprise main.

**Pull request merged**